src/librustc_mir/build/matches/simplify.rs

```diff
@@ -92,11 +92,22 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 Err(match_pair)
             }
 
-            PatternKind::Range { .. } |
-            PatternKind::Slice { .. } => {
+            PatternKind::Range { .. } => {
                 Err(match_pair)
             }
 
+            PatternKind::Slice { ref prefix, ref slice, ref suffix } => {
+                if prefix.is_empty() && slice.is_some() && suffix.is_empty() {
+                    // irrefutable
+                    candidate
+                        .match_pairs
+                        .push(MatchPair::new(match_pair.place, slice.as_ref().unwrap()));
```

Comment:
Huh, so, something here makes me a bit nervous. Why not invoke the helper `prefix_slice_suffix`? For example:


```rust
self.prefix_slice_suffix(
    &mut candidate.match_pairs,
    match_pair.place,
    prefix,
    slice,
    suffix,
);
```

that helper is more general than strictly needed, but I think it is sure to do the right thing, or at least we can do the wrong thing all in the same code. =)

Reply:
(In this particular case, the main difference would be that the `Place` would include a `Subslice` projection, I think, which may not be needed, but I still think i'd rather go through the helper.)

Reply:
Thoughts?

Reply:
In `prefix_slice_suffix()`, we have 
```rust
        if let Some(subslice_pat) = opt_slice {
            let subslice = place.clone().elem(ProjectionElem::Subslice {
                from: prefix.len() as u32,
                to: suffix.len() as u32
            });
            match_pairs.push(MatchPair::new(subslice, subslice_pat));
        }
```
so `ProjectionElem::Subslice.to` is set to `suffix.len()` which is `0` in this case.

Doc comment on `Subslice` in librustc/mir/mod.rs says:
```rust
    /// These indices are generated by slice patterns.
    ///
    /// slice[from:-to] in Python terms.
    Subslice {
        from: u32,
        to: u32,
    },
```
In Python, `slice[from:-0]` will be an empty list. This is why I avoided using `prefix_slice_suffix()`. 

However, it turns that setting `to` to `0` is actually equivalent to `slice[from:]` in Python term. So I agree that using `prefix_slice_suffix()` is better here.